src/igvfd/audit/file_set.py

```diff
@@ -21,3 +21,160 @@ def audit_no_files(value, system):
+        audit_category: missing seqspec
+        audit_levels: ERROR
+    '''
+    if 'files' in value:
+        no_seqspec = []
+        for file in value['files']:
+            if file.startswith('/sequence-files/'):
+                sequence_file_object = system.get('request').embed(file, '@@object?skip_calculated=true')
+                if not sequence_file_object.get('seqspec'):
+                    no_seqspec.append(file)
+        if no_seqspec:
+            no_seqspec = ', '.join([audit_link(path_to_text(file_no_seqspec), file_no_seqspec)
+                                   for file_no_seqspec in no_seqspec])
+            detail = (
+                f'File set {audit_link(path_to_text(value["@id"]), value["@id"])} has sequence file(s): '
+                f'{no_seqspec} which do not have a seqspec configuration file.'
+            )
+            yield AuditFailure('missing seqspec', detail, level='ERROR')
+
+
+@audit_checker('FileSet', frame='object')
+def audit_missing_seqspec_files(value, system):
+    '''
+        audit_detail: All files associated with a seqspec (both sequencing and seqspec files) are expected to be linked on the same file set.
+        audit_category: missing related seqspec files
+        audit_levels: ERROR
+    '''
+    if 'files' in value:
+        for file in value['files']:
+            if file.startswith('/sequence-files/'):
+                sequence_file_object = system.get('request').embed(file, '@@object?skip_calculated=true')
+
+                # Audit the file set with sequence files without the associated seqspec also in the file set.
+                if sequence_file_object.get('seqspec'):
+                    if sequence_file_object.get('seqspec') not in value['files']:
+                        seqspec_path = sequence_file_object['seqspec']
+                        detail = (
+                            f'File set {audit_link(path_to_text(value["@id"]), value["@id"])} has sequence file '
+                            f'{audit_link(path_to_text(file), file)} which links to seqspec '
+                            f'{audit_link(path_to_text(seqspec_path), seqspec_path)} which does not link to this file set.'
+                        )
+                        yield AuditFailure('missing related seqspec files', detail, level='ERROR')
+
+            # Audit the file set with a seqspec configuration file without the associated sequence files also in the file set.
+            if file.startswith('/configuration-files/'):
+                configuration_file_object = system.get('request').embed(file)
+                if configuration_file_object['content_type'] == 'seqspec' and configuration_file_object.get('seqspec_of'):
+                    missing_sequence_files = list(set(configuration_file_object.get(
+                        'seqspec_of', [])).difference(set(value['files'])))
+                    if missing_sequence_files:
```

Comment:
I am not sure that will work as expected, for example deleted files are not listed in the file_set['files'] and you would report them as linked to a different file_set

Reply:
Both seqspec_of and files filter for deleted files.